dial/throttler.go

```diff
@@ -0,0 +1,189 @@
+		fdLimit:   fdLimit,
+		peerLimit: perPeerLimit,
+
+		peerWaiters:   make(map[peer.ID][]*Job),
+		peerConsuming: make(map[peer.ID]int),
+		closeCh:       make(chan struct{}),
+
+		// we are capped by file descriptor limits, so it's safe to use as our chan lengths.
+		finished:       make(chan *Job, fdLimit),
+		fdTokenFreed:   make(chan *Job, fdLimit),
+		peerTokenFreed: make(chan *Job, fdLimit),
+	}
+}
+
+// NewThrottler returns a Throttler just like NewThrottlerWithParams, using default limits, governed by constants:
+//   * active peers per dial: 8
+//   * file descriptors: 160
+//
+// The LIBP2P_SWARM_FD_LIMIT can be used to override the file descriptors limit.
+func NewThrottler() Throttler {
+	fd := DefaultConcurrentFdDials
+	if env := os.Getenv("LIBP2P_SWARM_FD_LIMIT"); env != "" {
+		if n, err := strconv.ParseInt(env, 10, 32); err == nil {
+			fd = int(n)
+		}
+	}
+	return NewThrottlerWithParams(fd, DefaultPerPeerRateLimit)
+}
+
+func (t *throttler) Run(incoming <-chan *Job, released chan<- *Job) {
+	for {
+		// First, process token releases.
+		select {
+		case job := <-t.peerTokenFreed:
+			id := job.Request().PeerID()
+			t.peerConsuming[id]--
+			if t.peerConsuming[id] == 0 {
+				delete(t.peerConsuming, id)
+			}
+			job.Debugf("throttler: freeing peer token; waiting on peer limit: %d", len(t.peerWaiters[id]))
+			waitlist := t.peerWaiters[id]
+			if len(waitlist) == 0 {
+				continue
+			}
+			next := waitlist[0]
+			if len(waitlist) == 1 {
+				delete(t.peerWaiters, id)
+			} else {
+				waitlist[0] = nil
+				t.peerWaiters[id] = waitlist[1:]
+			}
+			if next.Cancelled() {
+				continue
+			}
+			t.peerConsuming[id]++
+			if !t.throttleFd(next) {
+				t.releaseJob(next, released)
+			}
+			continue
+
+		case job := <-t.fdTokenFreed:
+			t.fdConsuming--
+			job.Debugf("throttler: freeing FD token; waiting: %d; consuming: %d", len(t.fdWaiters), t.fdConsuming)
+			for len(t.fdWaiters) > 0 {
+				next := t.fdWaiters[0]
+				t.fdWaiters[0] = nil
+				t.fdWaiters = t.fdWaiters[1:]
+				if len(t.fdWaiters) == 0 {
+					t.fdWaiters = nil
+				}
+				if next.Cancelled() {
+					t.peerTokenFreed <- next
+					continue
+				}
+				t.fdConsuming++
+				t.releaseJob(next, released)
+				break
+			}
+
+		default:
+		}
+
+		// Next, make forward progress.
+		select {
+		case job := <-t.finished:
+			job.Debugf("throttler: clearing state for completed job")
+			if addrutil.IsFDCostlyTransport(job.Address()) {
+				t.fdTokenFreed <- job
+			}
+			t.peerTokenFreed <- job
+
+		case job := <-incoming:
+			if !t.throttlePeerLimit(job) && !t.throttleFd(job) {
+				t.releaseJob(job, released)
+			}
+
+		case <-t.closeCh:
+			return
+		}
+	}
+}
+
+func (t *throttler) releaseJob(job *Job, released chan<- *Job) {
+	job.MarkInflight()
+	job.AddCallback("throttler", func(j *Job) {
+		t.finished <- j
+	})
+	job.Debugf("throttler: releasing dial job to executor")
+	released <- job
+}
+
+func (t *throttler) throttleFd(job *Job) (throttle bool) {
```

Comment:
In future it might be nice to expose this (and other throttler info) to planners so e.g. we won't defer dials to relays and/or defer dialing expensive transports